src/Nethermind/Nethermind.AuRa/AuRaBlockFinalizationManager.cs

```diff
@@ -0,0 +1,269 @@
+            var minSealersForFinalization = block.IsGenesis ? 1 : _auRaValidator.MinSealersForFinalization;
+            var originalBlockSealer = block.Beneficiary;
+            
+            using (var batch = _blockInfoRepository.StartBatch())
+            {
+                var (chainLevel, blockInfo) = GetBlockInfo(block);
+                
+                while (!blockInfo.IsFinalized)
+                {
+                    validators.Add(block.Beneficiary);
+                    if (validators.Count >= minSealersForFinalization)
+                    {
+                        blockInfo.IsFinalized = true;
+                        _blockInfoRepository.PersistLevel(block.Number, chainLevel, batch);
+                        finalizedBlocks.Add(block);
+                    }
+
+                    if (!block.IsGenesis)
+                    {
+                        block = _blockTree.FindHeader(block.ParentHash, BlockTreeLookupOptions.None);
+                        (chainLevel, blockInfo) = GetBlockInfo(block);
+                    }
+                }
+            }
+
+            finalizedBlocks.Reverse(); // we were adding from the last to earliest, going through parents
+            
+            return finalizedBlocks;
+        }
+        */
+
+        public event EventHandler<FinalizeEventArgs> BlocksFinalized;
+
+        public long LastFinalizedBlockLevel
+        {
+            get => _lastFinalizedBlockLevel;
+            private set
+            {
+                if (_lastFinalizedBlockLevel < value)
+                {
+                    _lastFinalizedBlockLevel = value;
+                    if (_logger.IsTrace) _logger.Trace($"Setting {nameof(LastFinalizedBlockLevel)} to {value}.");
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            _blockProcessor.BlockProcessed -= OnBlockProcessed;
+        }
+
+        [DebuggerDisplay("Count = {Count}")]
+        private class ValidationStampCollection
+        {
+            private readonly ISet<Address> _set = new HashSet<Address>();
+            private readonly SortedList<long, Address> _list = new SortedList<long, Address>(Comparer<long>.Create((x, y) => (int) (y - x)));
+
+            public int Count => _set.Count;
+            
+            public void Add(BlockHeader blockHeader)
+            {
+                if (_set.Add(blockHeader.Beneficiary))
+                {
+                    _list.Add(blockHeader.Number, blockHeader.Beneficiary);
+                }
+            }
+
+            public void RemoveFromBlock(long blockNumber)
+            {
+                IEnumerable<long> toDelete = _list.Keys.SkipWhile(k => k > blockNumber).ToArray();
+                foreach (var number in toDelete)
+                {
+                    _set.Remove(_list[number]);
+                    _list.Remove(number);
+                }
+            }
+
+            public void Clear()
+            {
```

Comment:
have a look if you need resettables here (depending on the underlying collection growth charactersistics)

Reply:
by resettable I mean collections that implement IResettable and allow management of when to shrink / expand collection to avoid the the collections always growing to the peak levels and then not shrinking

Reply:
Discussed - no need as this should be fairly small, size of currently nbot finalized blocks.